apps/prairielearn/src/models/enrollment.ts

```diff
@@ -158,8 +228,157 @@ export async function generateAndEnrollUsers({
   return await runInTransactionAsync(async () => {
     const users = await generateUsers(count);
     for (const user of users) {
-      await ensureEnrollment({ course_instance_id, user_id: user.user_id });
+      await ensureEnrollment({
+        course_instance_id,
+        user_id: user.user_id,
+        // This is done by the system
+        agent_user_id: null,
+        agent_authn_user_id: null,
+        action_detail: 'implicit_joined',
+      });
     }
     return users;
   });
 }
+
+export async function selectEnrollmentById({ id }: { id: string }) {
+  return await queryRow(sql.select_enrollment_by_id, { id }, EnrollmentSchema);
+}
+
+/**
+ * Look up an enrollment by uid and course instance id.
+ * If there is no enrollment where the uid or pending_uid matches the given uid,
+ * this will return null.
+ */
+export async function selectOptionalEnrollmentByUid({
+  course_instance_id,
+  uid,
+}: {
+  course_instance_id: string;
+  uid: string;
+}) {
+  return await queryOptionalRow(
+    sql.select_enrollment_by_uid,
+    { course_instance_id, uid },
+    EnrollmentSchema,
+  );
+}
+
+/**
+ * This function invites an existing enrollment.
+ * All usages of this function should hold a lock on the enrollment.
+ * Callers should ensure that the enrollment is not already invited or joined.
+ */
+async function dangerouslyInviteExistingEnrollment({
+  enrollment_id,
+  agent_user_id,
+  pending_uid,
+  agent_authn_user_id,
+}: {
+  enrollment_id: string;
+  agent_user_id: string | null;
+  pending_uid: string;
+  agent_authn_user_id: string | null;
+}): Promise<Enrollment> {
+  const oldEnrollment = await selectEnrollmentById({ id: enrollment_id });
+
+  assert(oldEnrollment.status !== 'invited');
+  assert(oldEnrollment.status !== 'joined');
+
+  const newEnrollment = await queryRow(
+    sql.invite_existing_enrollment,
+    { enrollment_id, pending_uid },
+    EnrollmentSchema,
+  );
+
+  await insertAuditEvent({
+    table_name: 'enrollments',
+    action: 'update',
+    action_detail: 'invited',
+    row_id: newEnrollment.id,
+    old_row: oldEnrollment,
+    new_row: newEnrollment,
+    agent_user_id,
+    agent_authn_user_id,
+  });
+
+  return newEnrollment;
+}
+
+async function inviteNewEnrollment({
+  course_instance_id,
+  pending_uid,
+  agent_user_id,
+  agent_authn_user_id,
+}: {
+  course_instance_id: string;
+  pending_uid: string;
+  agent_user_id: string | null;
+  agent_authn_user_id: string | null;
+}) {
+  const newEnrollment = await queryRow(
+    sql.invite_new_enrollment,
+    { course_instance_id, pending_uid },
+    EnrollmentSchema,
+  );
+
+  await insertAuditEvent({
+    table_name: 'enrollments',
+    action: 'insert',
+    action_detail: 'invited',
+    row_id: newEnrollment.id,
+    new_row: newEnrollment,
+    subject_user_id: null,
+    agent_user_id,
+    agent_authn_user_id,
+  });
+
+  return newEnrollment;
+}
+/**
+ * Invite a student by uid. This does not prevent being invited via
```

Comment:
Incomplete comment?

Reply:
Removed the incomplete part in c0add6bacf9a32aed633d9d75e2ffd537bfce788.